File: ts/31.next-permutation.ts
/*
 * @lc app=leetcode id=31 lang=typescript
 *
 * [31] Next Permutation
 */

// @lc code=start
/**
 Do not return anything, modify nums in-place instead.
 */
function nextPermutation(nums: number[]): void {
  const reverse = (start: number) => {
    let i = start;
    let j = nums.length - 1;

    while (i < j) {
      swap(i, j);
      i++;
      j--;
    }
  };

  const swap = (i: number, j: number) => {
    let temp = nums[i];
    nums[i] = nums[j];
    nums[j] = temp;
  };

  let i = nums.length - 2;
  while (i >= 0 && nums[i + 1] <= nums[i]) i--;

  if (i >= 0) {
    let j = nums.length - 1;
    while (nums[j] <= nums[i]) j--;
    swap(i, j);
  }

  reverse(i + 1);
}
// @lc code=end

File: go/31.next-permutation.go
/*
 * @lc app=leetcode id=31 lang=golang
 *
 * [31] Next Permutation
 */
package leetcode

// @lc code=start
func nextPermutation(nums []int) {
	n := len(nums)

	if n == 1 {
		return
	}

	l, r := n-2, n-1

	for l >= 0 && nums[l] >= nums[l+1] {
		l--
	}

	if l >= 0 {
		for nums[r] <= nums[l] {
			r--
		}
		swap(&nums, l, r)
	}

	reverse(&nums, l+1)
}

func swap(nums *[]int, i int, j int) {
	(*nums)[i], (*nums)[j] = (*nums)[j], (*nums)[i]
}

func reverse(nums *[]int, start int) {
	for i, j := start, len(*nums)-1; i < j; i, j = i+1, j-1 {
		swap(nums, i, j)
	}
}

// @lc code=end
